app/coffee/UpdatesManager.coffee

```diff
@@ -70,51 +70,79 @@ module.exports = UpdatesManager =
-				# get the updates as strings from redis (so we can delete them after they are applied)
-				RedisManager.getOldestDocUpdates doc_id, UpdatesManager.REDIS_READ_BATCH_SIZE, (error, docUpdates) ->
+				logger.log project_id: project_id, doc_id: doc_id, rawUpdates: rawUpdates, "retrieved raw updates from redis"
+				UpdatesManager.compressAndSaveRawUpdates project_id, doc_id, rawUpdates, temporary, (error) ->
 					return callback(error) if error?
-					length = docUpdates.length
-					# parse the redis strings into ShareJs updates
-					RedisManager.expandDocUpdates docUpdates, (error, rawUpdates) ->
+					logger.log project_id: project_id, doc_id: doc_id, "compressed and saved doc updates"
+					# delete the applied updates from redis
+					RedisManager.deleteAppliedDocUpdates project_id, doc_id, docUpdates, (error) ->
 						return callback(error) if error?
-						logger.log project_id: project_id, doc_id: doc_id, rawUpdates: rawUpdates, "retrieved raw updates from redis"
-						UpdatesManager.compressAndSaveRawUpdates project_id, doc_id, rawUpdates, temporary, (error) ->
-							return callback(error) if error?
-							logger.log project_id: project_id, doc_id: doc_id, "compressed and saved doc updates"
-							# delete the applied updates from redis
-							RedisManager.deleteAppliedDocUpdates project_id, doc_id, docUpdates, (error) ->
-								return callback(error) if error?
-								if length == UpdatesManager.REDIS_READ_BATCH_SIZE
-									# There might be more updates
-									logger.log project_id: project_id, doc_id: doc_id, "continuing processing updates"
-									setTimeout () ->
-										UpdatesManager.processUncompressedUpdates project_id, doc_id, callback
-									, 0
-								else
-									logger.log project_id: project_id, doc_id: doc_id, "all raw updates processed"
-									callback()
+						if length == UpdatesManager.REDIS_READ_BATCH_SIZE
+							# There might be more updates
+							logger.log project_id: project_id, doc_id: doc_id, "continuing processing updates"
+							setTimeout () ->
+								UpdatesManager._processUncompressedUpdates project_id, doc_id, temporary, callback
+							, 0
+						else
+							logger.log project_id: project_id, doc_id: doc_id, "all raw updates processed"
+							callback()
 
+	# Process updates for a doc when we flush it individually
 	processUncompressedUpdatesWithLock: (project_id, doc_id, callback = (error) ->) ->
-		LockManager.runWithLock(
-			"HistoryLock:#{doc_id}",
-			(releaseLock) ->
-				UpdatesManager.processUncompressedUpdates project_id, doc_id, releaseLock
-			callback
-		)
+		UpdatesManager._prepareProjectForUpdates project_id, (error, temporary) ->
+			return callback(error) if error?
+			UpdatesManager._processUncompressedUpdatesForDoc project_id, doc_id, temporary, callback
+
+
+	# Process updates for a doc when the whole project is flushed (internal method)
+	_processUncompressedUpdatesForDoc: (project_id, doc_id, temporary, callback = (error) ->) ->
```

Comment:
This should be called `_processUncompressedUpdatesForDocWithLock` since it is responsible for taking the lock.